cpp/src/decisiontree/decisiontree.cuh

```diff
@@ -0,0 +1,425 @@
+                             tl_tree.RightChild(node_id));
+        ++next_end;
+
+        // Set node from current level as numerical node. Children IDs known.
+        tl_tree.SetNumericalSplit(node_id, q_node.node->colid,
+                                  q_node.node->quesval, true,
+                                  tl::Operator::kLE);
+
+      } else {
+        if (num_class == 1) {
+          tl_tree.SetLeaf(node_id, static_cast<T>(q_node.node->prediction));
+        } else {
+          std::vector<T> leaf_vector(num_class, 0);
+          leaf_vector[q_node.node->prediction] = 1;
+          tl_tree.SetLeafVector(node_id, leaf_vector);
+        }
+      }
+    }
+
+    cur_level_queue.swap(next_level_queue);
+    cur_front = next_front;
+    cur_end = next_end;
+    next_front = 0;
+    next_end = 0;
+  }
+  return tl_tree;
+}
+
+struct DataInfo {
+  unsigned int NLocalrows;
+  unsigned int NGlobalrows;
+  unsigned int Ncols;
+};
+
+template <class T, class L>
+class DecisionTree {
+ protected:
+  DataInfo dinfo;
+  int depth_counter = 0;
+  int leaf_counter = 0;
+  int n_unique_labels = -1;  // number of unique labels in dataset
+  double prepare_time = 0;
+  double train_time = 0;
+  MLCommon::TimerCPU prepare_fit_timer;
+  DecisionTreeParams tree_params;
+
+ public:
+  /**
+     * @brief Fits a DecisionTree on given input data and labels
+     * @param[in] handle             cuML handle
+     * @param[in] data               pointer to input training data
+     * @param[in] ncols              number of features (columns)
+     * @param[in] nrows              number of samples (rows)
+     * @param[in] labels             pointer to label data
+     * @param[in] rowids             pointer to array of row indices mapping to data
+     * @param[in] n_sampled_rows     count of rows sampled
+     * @param[in] unique_labels      count of unique labels
+     * @param[in] is_classifier      true if task is classification, else false
+     * @param[in,out] tree           pointer to tree structure
+     * @param[in] tree_parameters    structure of tree parameters
+     * @param[in] seed               random seed
+     * @param[in] d_global_quantiles device pointer to global quantiles
+     */
+  void fit(const raft::handle_t &handle, const T *data, const int ncols,
+           const int nrows, const L *labels, unsigned int *rowids,
+           const int n_sampled_rows, int unique_labels, bool is_classifier,
+           DT::TreeMetaDataNode<T, L> *&tree,
+           DecisionTreeParams tree_parameters, uint64_t seed,
+           T *d_global_quantiles) {
+    this->tree_params = tree_parameters;
+    this->prepare_fit_timer.reset();
+    const char *CRITERION_NAME[] = {"GINI", "ENTROPY", "MSE", "MAE", "END"};
+    CRITERION default_criterion =
+      (is_classifier) ? CRITERION::GINI : CRITERION::MSE;
+    CRITERION last_criterion =
+      (is_classifier) ? CRITERION::ENTROPY : CRITERION::MSE;
+
+    validity_check(tree_params);
```

Comment:
Can we check that the label type is consistent with the criterion somewhere?

Reply:
I initially thought of resolving the ambiguity with `unique_labels` (1 for regression (as not meaningful for classification), >1 for clf). What do you think?

Reply:
I'm not sure if that's reliable. One way of doing this would be to add the method `bool CheckLabel(LabelT label)` to each objective, and then do `thrust::all_of` on the labels somewhere. That way each objective can specify the input it expects.

Reply:
Actually, I don't think we need the `thrust::all_of` because we don't need to check the numerical values only the type with something like std::is_integer.